Syn decode in publisher

DaemonSnake opened the pull request:
move decoding of raw binary messages out of the hot-path
might help reduce walex lag

DaemonSnake committed:
server speed up: allow decoding to be done in publisher

Experiment to check if moving decoding in Publisher and outside the main
connection server allows to increase its throughput

DaemonSnake committed:
fix small type violation

error:

    warning: incompatible types given to Kernel.hd/1:

        hd(__CALLER__.context_modules)

    given types:

        list(atom())

    but expected one of:

        non_empty_list(term(), term())

    typing violation found at:
    │
  3 │     module = hd(__CALLER__.context_modules)
    │              ~
    │
    └─ lib/walex/event/dsl.ex:3:14: WalEx.Event.Dsl.process_events_async/2

MrYawe commented on lib/walex/event/dsl.ex:
```diff
@@ -1,6 +1,7 @@
 defmodule WalEx.Event.Dsl do
   defmacro process_events_async(events, functions) do
-    module = hd(__CALLER__.context_modules)
+    [_ | _] = __CALLER__.context_modules
```
Is this line useful?

DaemonSnake replied:
yes, context_modules can be an empty list according to its type
if so this macro should crash
List.first returns a nil value in that case

DaemonSnake committed:
process: set :message_queue_data -> :off_heap in all GenServer that can receive many messages

following erlang doc:

> If the process may potentially accumulate a large number of messages in its queue it is recommended to set the flag value to off_heap.

DaemonSnake commented:
decoding in the publisher didn't seem to bring much to the table
message_queue_data: :off_heap was merged in walex's latest relase
Closing this PR
